internal/native/grpc_client.go

```diff
@@ -0,0 +1,258 @@
+	client pb.NativeServiceClient
+	logger *zerolog.Logger
+
+	eventStream pb.NativeService_StreamEventsClient
+	eventM      sync.RWMutex
+	eventCh     chan *pb.Event
+	eventDone   chan struct{}
+
+	onVideoStateChange func(state VideoState)
+	onIndevEvent       func(event string)
+	onRpcEvent         func(event string)
+
+	closed bool
+	closeM sync.Mutex
+}
+
+type grpcClientOptions struct {
+	SocketPath         string
+	Logger             *zerolog.Logger
+	OnVideoStateChange func(state VideoState)
+	OnIndevEvent       func(event string)
+	OnRpcEvent         func(event string)
+}
+
+// NewGRPCClient creates a new gRPC client connected to the native service
+func NewGRPCClient(opts grpcClientOptions) (*GRPCClient, error) {
+	// Connect to the Unix domain socket
+	conn, err := grpc.NewClient(
+		fmt.Sprintf("unix-abstract:%v", opts.SocketPath),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to gRPC server: %w", err)
+	}
+
+	client := pb.NewNativeServiceClient(conn)
+
+	grpcClient := &GRPCClient{
+		conn:               conn,
+		client:             client,
+		logger:             opts.Logger,
+		eventCh:            make(chan *pb.Event, 100),
+		eventDone:          make(chan struct{}),
+		onVideoStateChange: opts.OnVideoStateChange,
+		onIndevEvent:       opts.OnIndevEvent,
+		onRpcEvent:         opts.OnRpcEvent,
+	}
+
+	// Start event stream
+	go grpcClient.startEventStream()
+
+	return grpcClient, nil
+}
+
+func (c *GRPCClient) setStream(stream pb.NativeService_StreamEventsClient) {
+	c.eventM.Lock()
+	defer c.eventM.Unlock()
+	c.eventStream = stream
+}
+
+func (c *GRPCClient) handleEventStream(stream pb.NativeService_StreamEventsClient) {
+	logger := *c.logger
+	for {
+		if stream == nil {
+			logger.Error().Msg("event stream is nil")
+			break
+		}
+
+		event, err := stream.Recv()
+
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				logger.Debug().Msg("event stream closed")
+			} else {
+				logger.Warn().Err(err).Msg("event stream error")
+			}
+			break
+		}
+
+		// enrich the logger with the event type and data, if debug mode is enabled
+		if c.logger.GetLevel() <= zerolog.DebugLevel {
+			logger = logger.With().
+				Str("type", event.Type).
+				Interface("data", event.Data).
+				Logger()
+		}
+		logger.Trace().Msg("received event")
+
+		select {
+		case c.eventCh <- event:
+		default:
+			logger.Warn().Msg("event channel full, dropping event")
+		}
+	}
+}
+
+func (c *GRPCClient) startEventStream() {
+	for {
+		// check if the client is closed
+		c.closeM.Lock()
+		if c.closed {
+			c.closeM.Unlock()
+			return
+		}
+		c.closeM.Unlock()
+
+		ctx := context.Background()
+		stream, err := c.client.StreamEvents(ctx, &pb.Empty{})
+		if err != nil {
+			c.logger.Warn().Err(err).Msg("failed to start event stream, retrying ...")
+			time.Sleep(5 * time.Second)
+			continue
+		}
+
+		c.setStream(stream)
+		c.handleEventStream(stream)
+		c.setStream(nil)
```

Comment:
Seems like a good place for a defer :) to nil the stream... 

Maybe it would be clearer to have the handleEventStream set and nil it itself?